### ISSUE TITLE
Tracer Correlation gets reset when using YARP

### ISSUE THREAD

**Issue by zikhan** [DataDog/dd-trace-dotnet#3832]
**Describe the bug**
I've implemented an extremely simple reverse proxy using YARP's [Direct Forwarder](https://microsoft.github.io/reverse-proxy/articles/direct-forwarding.html). I've noticed that the `x-datadog-trace-id` is coming into the request fine, however it looks like it creates an entirely different APM trace for the call in to the service. The header logging can be seen in the code screenshot. I am using the minimal api hosting model.

I see the calling http-client in a trace.
I see the receiving service in a different trace.

If the request is fresh with no correlation information, then everything works as it should.

UPDATE:
I tried downgrading to .NET 6, ensuring the SocketsHttpHandler integration was being used, ensuring that the upstream services were on the same version of the trace package, ensured the opt-in from https://github.com/DataDog/dd-trace-dotnet/issues/668#issuecomment-616805749 was being used. All of it produced the same behavior. 

**Expected behavior**
Traces should be correlated.

**Screenshots**
RouterMiddleware
![Image of the middleware doing the work](https://user-images.githubusercontent.com/8488326/221273363-ffa48d51-e87b-46b5-b997-4fe69d0a26c1.png)

Program.cs
![Image of Program.cs](https://user-images.githubusercontent.com/8488326/221275211-e4096a01-457f-4334-b004-d07018ae0a52.png)

**Runtime environment (please complete the following information):**
- Instrumentation mode: Automatic in Container
- Tracer version: 2.24.0
- OS: alpine 3.17
- CLR: .net 7

**Additional context**
I will have an issue getting the tracer logs since the traced environment is in a k8s cluster that I don't have immediate access to.

<!--
NOTE: To contact our support team, see https://www.datadoghq.com/support/
When contacting support, please include the information requested above.
If possible, also include the tracer log files found (by default) in:
Windows: %ProgramData%\Datadog .NET Tracer\logs\
Linux: /var/log/datadog/dotnet/
--!>

**Comment by zikhan** [comment]
Not a problem! Thanks for taking the time!

Requested Information:
* DataDog Configuration
  <details><summary>In DockerFile</summary>
  
  ```Dockerfile
  ENV DD_INTEGRATIONS=/opt/datadog/integrations.json
  ENV DD_TRACE_AGENT_HOSTNAME=datadog-agent
  ENV DD_TRACE_AGENT_PORT=8126
  ENV DD_LOGS_INJECTION=true
  ENV DD_TRACE_ANALYTICS_ENABLED=true
  ENV DD_RUNTIME_METRICS_ENABLED=true
  ENV DD_DOTNET_TRACER_HOME=/opt/datadog
  ENV DD_SERVICE=court-proxy
  ENV DD_TRACE_HEADER_TAGS=X-Tenant-Id:tenantId
  ```
  
  </details>
  <details><summary>In K8s deployment</summary>

  <p>
  DD_AGENT_HOST= &lt;Referenced host ip&gt;<br>
  DD_VERSION=&lt;Application Version&gt;<br>
  DD_TAGS="division:division,product:product,customer-id:multi"<br>
  <br>
  Label on the pod:<br>
  tags.datadoghq.com/service:court-proxy
  </p>
  </details>

  DD Environment is coming from the dd agent in the k8s cluster

* No Manual tracing. NuGet package isn't even referenced.
* Here's an example where a new trace id was created but the http request log shows a different Id
  HttpRequestLog (I did not see a trace state header)
  ![image](https://user-images.githubusercontent.com/8488326/225381574-558d0987-652e-4276-9561-18ff228619b3.png)
  APM Span
  ![image](https://user-images.githubusercontent.com/8488326/225382283-55a0cadf-4084-49f9-b944-bd2047faf382.png)

You can see I found the http request log was enriched with a completely different traceid, even though the incoming request header had the correct upstream trace id.

Some other weird behavior I just noticed has I was gathering the screenshots; it appears that these spans are sharing traceid's for other requests being sent from the upstream service. In that span above, all of the spans were correlated together as if they were tied to the correct trace, however none of them were linked to the upstream trace.

That incoming x-datadog-trace-id on the http request log is a correct trace. Here's the upstream trace that the spans are supposed to correlate to:
![image](https://user-images.githubusercontent.com/8488326/225381975-03b38f7c-7681-4493-9a52-2b930e587725.png)

Some other info that might help, the upstream services are running tracer library version: 2.20. I updated to 2.26 but that didn't seem to help either.

**Comment by andrewlock** [comment]
Hi @zikhan,

Thanks, that's really helpful, and goes part way to explaining things.

The tl;dr; is that the `traceparent` and `x-datadog-trace-id` in your requests are providing different values. 

**Looking at the injected values**

If we take the `traceparent` from that request, `00-0a2a0c59c2587e265bba634e1b9bdfbb-539116310397210c-00`, and run it through our parser, we get the following values:

- `TraceId: 6609704590241423291`
- `ParentId: 6021618576540508428`

Those values are clearly incompatible with the other header values in the request:

- `x-datadog-trace-id: 8329154629666547108`
- `x-datadog-parent-id: 6580157064720572566`

**Why is it a problem?**

In version 2.22.0 of the tracer ([in this PR](https://github.com/DataDog/dd-trace-dotnet/pull/3583)), we switched to using the `traceparent` header preferentially for propagation. That explains the behaviour you're seeing in the downstream service:

- The upstream service injects a `traceparent` that differs from the `x-datadog-trace-id` (we'll come to that shortly)
- You're using tracer version 2.22.0+ in the downstream service
- It sees the `traceparent` and `x-datadog-trace-id` headers
- It preferentially uses the `traceparent` values
- Your trace ends up disconnected.

So that brings us to the main question: _why does the injected `traceparent` differ from the `x-datadog-trace-id`_?

**Why do we have incompatible values?**

I can't say for sure based on the screenshots provided so far, but I can make one specific inference:

- The `traceparent` header is _not_ being injected by the .NET tracer library. I can tell that because it's using a 128bit trace ID, whereas we currently only support 64bit (128 bit support is actively being worked on, and we're _compatible_ with 128bit, it just means that _we_ didn't inject that value).

You said that you still had the same issue after updating the "upstream" services to 2.26.0? This is where terminology differs, do you mean you updated the _proxy_, the _backend_ service, or the app that _calls_ the proxy to 2.26.0? There's a lot of moving pieces 😅 

Prior to 2.22.0, we _only_ injected the `x-datadog-trace-id` headers, _not_ `traceparent`, into the outgoing requests. If there is a component in the stack that doesn't understand the datadog headers, but _does_ understand `traceparent`, that component might inject a new `traceparent` that differs from our injected `x-datadog-trace-id`. 

Similarly, if a service is instrumented with a version of the Datadog tracer that _doesn't_ understand `traceparent` (i.e. < 2.22.0 for the .NET tracer, different for other languages), then the tracer library may ignore an incoming `traceparent` value and inject the `x-datadog-trace-id` value instead. If the framework automatically propagates `traceparent`, then you will have a similar mismatch.

In your screenshot, I see various intermediate services, but my _guess_ would be that the problem happens in the `court.shell` service which, AFAICT, calls the `court-proxy` service (which has the trace id discrepency). Can you confirm what version of the tracer is running in that service?

**How can you fix it?**

There's basically 2 ways to fix this currently:
- Ensure the `court.shell` is running a version of the .NET tracer that supports `traceparent` (v2.22.0+, ideally the latest, as there are bug fixes for the support)
- Disable (for now) `traceparent` extraction in the `court-proxy` service.

The former will be the "best" way long term. If you can upgrade all your tracer versions to versions that support `traceparent`, you can rely on the w3c standard headers, and everything should work as you expect, including interoperating with other components which _only_ support `traceparent` (and not the `x-datadog-*` headers.

The latter option is a stop-gap while services are sending incompatible headers. You can set the supported propagation styles as described [in the docs](https://docs.datadoghq.com/tracing/trace_collection/library_config/dotnet-core?tab=environmentvariables#headers-extraction-and-injection), using the `DD_TRACE_PROPAGATION_STYLE_INJECT` and `DD_TRACE_PROPAGATION_STYLE_EXTRACT` headers. For example, if you set `DD_TRACE_PROPAGATION_STYLE_EXTRACT=Datadog`, that will disable the extraction of `traceparent` and should solve your issue. 

Note that you can also have different values for injection and extraction. So you could disable extraction of `traceparent` in this service, which would fix your issue, but continue to inject `traceparent` for downstream services. They would then get a "new" `traceparent` value that matches the injected `x-datadog-*` headers too. That might mean you only need to "fix" a couple of services until support for `traceparent` is added to the upstream services.

Sorry for the excessive explanation (it's my curse), but I wanted you to have the full picture! 😅 I hope it makes sense and that you can resolve your issue. Any questions let me know!

cc @lucaspimentel as you might have more to add! Also, should we [update the docs](https://docs.datadoghq.com/tracing/trace_collection/library_config/dotnet-core?tab=environmentvariables#headers-extraction-and-injection) as I don't think it's correct about the default values anymore?

**Comment by zikhan** [comment]
I'll take the long explanation any day! This is awesome and it'll be good to be aware of the different propagation styles, as I only had the `x-datadog-*` one's in my head. I've gone ahead and done the second option which works in the environment as it stands today. When we get to updating all the upstream services, I'll switch back to the default. But now I know what to look for! Appreciate the help!
